text/0000-variadic-generics.md

```diff
@@ -0,0 +1,209 @@
+demonstrates that Rust's lack of variadic generics is resulting in a subpar
+edit-compile-debug cycle for at least one prominent, high-quality crate.
+
+The solution proposed here would resolve the limitations above by making it
+possible to implement traits for tuples of arbitrary length. This change would
+make Rust libraries easier to understand and improve the edit-compile-debug
+cycle when using variadic code.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## The `Tuple` Trait
+The following would be implemented by all tuple types:
+```rust
+trait Tuple {
+    type AsRefs<'a>: Tuple + 'a;
+    type AsMuts<'a>: Tuple + 'a;
+    fn elements_as_refs<'a>(&'a self) -> Self::AsRefs<'a>;
+    fn elements_as_mut<'a>(&'a mut self) -> Self::AsMuts<'a>;
+}
+```
+
+The types `AsRefs` and `AsMuts` are the corresponding tuples of references to
+each element in the original tuple. For example,
+`(A, B, C)::AsRefs = (&A, &B, &C)` and
+`(A, B, C)::AsMuts = (&mut A, &mut B, &mut C)`
+
+The `Tuple` trait should only be implemented for tuples and marked with the
+`#[fundamental]` attribute described in
+[the coherence RFC](https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md).
+This would allow coherence and type-checking to be extended to assume that no
+implementations of `Tuple` will be added. This enables an increased level of
+negative reasoning making it easier to write blanket implementations of traits
+for tuples.
+
+## The `(Head, ...Tail)` Type Syntax
+This syntax would allow for a `Cons`-cell-like representation of tuple types.
+For example, `(A, ...(B, C))` would be equivalent to `(A, B, C)`. This allows
+users to represent the type of tuples in an inductive style when writing trait
+implementations.
+
+## The `(head, ...tail)` Pattern-Matching Syntax
+This syntax allows for splitting apart the head and tail of a tuple. For
+example, `let (head, ...tail) = (1, 2, 3);` moves the head value, `1`, into
+`head`, and the tail value, `(2, 3)`, into `tail`.
+
+## The `(head, ...tail)` Joining Syntax
+This syntax allows pushing an element onto a tuple. It is the natural inverse
+of the pattern-matching operation above. For example,
+`let tuple = (1, ...(2, 3));` would result in `tuple` having a value of
+`(1, 2, 3)`.
+
+## An Example
+
+Using the tools defined above, it is possible to implement `TupleMap`, a
+trait which can apply a mapping function over all elements of a tuple:
+
+```rust
+trait TupleMap<F>: Tuple {
+    type Out: Tuple;
+    fn map(self, f: F) -> Self::Out;
+}
+
+impl<F> TupleMap<F> for () {
+    type Out = ();
+    fn map(self, _: F) {}
+}
+
+impl<Head, Tail, F, R> TupleMap<F> for (Head, ...Tail)
+    where
+    F: Fn(Head) -> R,
+    Tail: TupleMap<F>,
+{
+    type Out = (R, ...<Tail as TupleMap<F>>::Out);
+    
+    fn map(self, f: F) -> Self::Out {
+        let (head, ...tail) = self;
+        let mapped_head = f(head);
+        let mapped_tail = tail.map(f);
+        (mapped_head, ...mapped_tail)
```

Comment:
Might be a bit clearer with the last 3 lines flattened into `(f(head), ...tail.map(f))`.